### ISSUE TITLE
Trilinos nightly configuration fails with Kokkos' develop branch

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#6152]
**Describe the bug**
The nightly Trilinos build, Serial backend, with clang/10 began failing to configure with output:
```
CMake Error at kokkos/cmake/kokkos_enable_options.cmake:37 (KOKKOS_ENABLE_OPTION):
  KOKKOS_ENABLE_OPTION Function invoked with incorrect arguments for function
  named: KOKKOS_ENABLE_OPTION
Call Stack (most recent call first):
  kokkos/cmake/kokkos_tribits.cmake:201 (INCLUDE)
  kokkos/CMakeLists.txt:226 (KOKKOS_SETUP_BUILD_ENVIRONMENT)


CMake Error at kokkos/cmake/kokkos_enable_options.cmake:113 (MESSAGE):
  Internal logic error: option 'CUDA_LAMBDA' or device 'CUDA' not recognized.
Call Stack (most recent call first):
  kokkos/cmake/kokkos_enable_options.cmake:123 (CHECK_DEVICE_SPECIFIC_OPTIONS)
  kokkos/cmake/kokkos_tribits.cmake:201 (INCLUDE)
  kokkos/CMakeLists.txt:226 (KOKKOS_SETUP_BUILD_ENVIRONMENT)
```

This was with a Serial build, no configure options involving the CUDA backend (in particular Kokkos_ENABLE_CUDA_LAMBDA) were passed to the cmake configure line

This began after merge of the following commits on May 19:
Only pass one wrapper object in SYCL reductions ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))
    Explicitly cast to CombinedFunctorReducerType ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))
    Expand list of kokkos options not to export with cmake ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))
    Do not append to Kokkos_OPTIONS variables those in the do not export list ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))
    Drop Kokkos_ENABLE_LAUNCH_COMPILER option ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))
    Export Kokkos_ENABLE_<OPTION> that are relevant ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))
    Suppress bogus warning about CUDA_LAMBDA being ON ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))
    [ci skip] Add nightly ci for spack (#6135) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))
    OpenMPTarget: Enable Cray compiler for the OpenMPTarget backend. (#5889) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Clang1001_Serial_NoMPI_debug-develop/122/changes#detail))

**Please also include the following items to support reproducing the bug**
**Reproducer** (Blake testbed, SKX arch, clang/10 Serial build):
```bash
# Repos
git clone -b develop https://github.com/trilinos/Trilinos.git
git clone -b develop https://github.com/kokkos/kokkos.git
git clone -b develop https://github.com/kokkos/kokkos-kernels.git
# Symbolic link to your kokkos and kokkos-kernels repos in Trilinos source directory for source override
cd Trilinos
ln -s <path-to-your-repo>/kokkos kokkos
ln -s <path-to-your-repo>/kokkos-kernels kokkos-kernels

cd $HOME
mkdir -p build
cd build

module load cmake/3.25.2 clang/10.0.1 openblas/0.3.13/gcc/10.2.0

cmake \
 -D CMAKE_INSTALL_PREFIX="${TRILINOS_INSTALL_DIR}" \
 -D CMAKE_CXX_STANDARD="17" \
 -D CMAKE_CXX_FLAGS="-g" \
 -D CMAKE_BUILD_TYPE=DEBUG \
 -D BUILD_SHARED_LIBS:BOOL=ON \
\
 -D TPL_ENABLE_MPI=OFF \
\
 -D TPL_ENABLE_BLAS:STRING=ON \
  -D BLAS_LIBRARY_DIRS:FILEPATH=${BLAS_ROOT}/lib \
  -D BLAS_LIBRARY_NAMES:STRING="openblas" \
 -D TPL_ENABLE_LAPACK:STRING=ON \
  -D LAPACK_INCLUDE_DIRS:FILEPATH="${LAPACK_ROOT}/include" \
  -D LAPACK_LIBRARY_DIRS:FILEPATH=${LAPACK_ROOT}/lib \
  -D LAPACK_LIBRARY_NAMES:STRING="openblas" \
\
 -D Trilinos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_EXAMPLES=ON \
 -D Trilinos_ENABLE_COMPLEX=ON \
\
 -D Trilinos_ENABLE_Ifpack2=ON \
  -D Ifpack2_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Amesos2=ON \
  -D Amesos2_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Kokkos=ON \
 -D Kokkos_ENABLE_SERIAL=ON \
 -D Kokkos_ARCH_SKX=ON \
  -D Kokkos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_KokkosKernels=ON \
  -D KokkosKernels_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Tpetra=ON \
  -D Tpetra_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Sacado=ON \
  -D Sacado_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Stokhos=ON \
  -D Stokhos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Zoltan2=ON \
  -D Zoltan2_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Intrepid2=ON \
  -D Intrepid2_ENABLE_TESTS=OFF \
 -D Trilinos_ENABLE_Belos=ON \
  -D Belos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Anasazi=ON \
  -D Anasazi_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Teuchos=ON \
  -D Teuchos_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_MueLu=ON \
  -D MueLu_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Panzer=ON \
  -D Panzer_ENABLE_TESTS=ON \
 -D Trilinos_ENABLE_Phalanx=ON \
  -D Phalanx_ENABLE_TESTS=ON \
 -D Kokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
 -D KokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
$TRILINOS_DIR
```

**Comment by masterleinad** [comment]
Oh, `Trilinos` doesn't specify a value for `Kokkos_ENABLE_CUDA` and that's why we see the configure-time error there.

**Comment by masterleinad** [comment]
We previously had https://github.com/kokkos/kokkos/blob/62d2b6c879b74b6ae7bd06eb3e5e80139c4708e6/cmake/kokkos_enable_options.cmake#L75-L82
so we should probably also set it to `OFF` for `Trilinos_ENABLE_Kokkos AND NOT TPL_ENABLE_CUDA`

**Comment by ndellingwood** [comment]
@masterleinad should the local var `CUDA_LAMBDA_DEFAULT` and logic you posted be reintroduced, or would you prefer something like below, or something else:

```cmake
IF (Trilinos_ENABLE_Kokkos AND NOT TPL_ENABLE_CUDA) 
  KOKKOS_ENABLE_OPTION(CUDA_LAMBDA OFF "Whether to allow lambda expressions on the device with NVCC **DEPRECATED**") 
ELSE()
  KOKKOS_ENABLE_OPTION(CUDA_LAMBDA ${Kokkos_ENABLE_CUDA} "Whether to allow lambda expressions on the device with NVCC **DEPRECATED**")
ENDIF()
```

**Comment by masterleinad** [comment]
We should probably just reintroduce `CUDA_LAMBDA_DEFAULT` like we had before. After all, we didn't change anything for the option itself anyway (apart from the doc string and introducing some warnings/errors if it wasn't set).

**Comment by dalg24** [comment]
I would rather fix https://github.com/kokkos/kokkos/blob/382f0bea7bfe2228716dcd89c15b3a5cb2c0833a/cmake/kokkos_enable_options.cmake#L123

**Comment by masterleinad** [comment]
The first error comes from defining the `CMake` option itself so we need to do something about that anyway (and depending on that solution possibly something for `CHECK_DEVICE_SPECIFIC_OPTIONS`).

**Comment by dalg24** [comment]
Can you please elaborate here.  The first error seems to be that `Kokkos_ENABLE_CUDA` is not defined when 
https://github.com/kokkos/kokkos/blob/5d81422daea73f5a2a69771cc0dfafc19f785003/cmake/kokkos_enable_options.cmake#L37

What is different in Trilinos?

**Comment by masterleinad** [comment]
> What is different in Trilinos?

The description reads
```
//Enable optional library support in the package Kokkos for the
// package CUDA.  Set to 'ON', 'OFF', or leave empty to allow for
// other logic to decide.
Kokkos_ENABLE_CUDA:STRING=
```
which comes from https://github.com/trilinos/Trilinos/blob/d9fb605c3fa6d312513bc51eeadbc6c1e9c55e3e/cmake/tribits/core/package_arch/TribitsAdjustPackageEnables.cmake#L1215-L1239.

**Comment by masterleinad** [comment]
Effectively `TriBITS` parses TPLs before configuring projects and defines corresponding `CMake` options before we have a chance to do that.

**Comment by ndellingwood** [comment]
Resolved by #6156

**Comment by masterleinad** [comment]
Fixed in #6156.
